### PR TITLE
Sparse hessian with Symbolics.jl

### PR DESCRIPTION
**Pull request by amontoison**

### CODE REVIEW

File: src/sparse_jacobian.jl [hunk 1]
```diff
@@ -111,7 +112,7 @@ end
 
 function jac_structure!(
   b::SparseADJacobian,
-  ::ADModel,
+  nlp::ADModel,
```
Comment (tmigot) [inline src/sparse_jacobian.jl:115]: Is there a reason to modify this?

**Review by tmigot — CHANGES_REQUESTED** [review]
Thanks @amontoison for the PR that looks great overall.

**Comment by tmigot** [comment]
Ok, so I didn't check the PR in the right order, some comments were valid for https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/125/files

File: src/sparse_hessian.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  @variables xs[1:nvar], μs
+  xsi = Symbolics.scalarize(xs)
+  fun = μs * f(xsi)
+  if ncon > 0
+    @variables ys[1:ncon]
+    ysi = Symbolics.scalarize(ys)
+    cx = similar(ysi)
+    fun = fun + dot(c!(cx,xsi), ysi)
+  end
+  S = Symbolics.hessian_sparsity(fun, ncon == 0 ? xsi : [xsi; ysi], full=false)
```
[the remark this answers is not in the document]
Comment (amontoison) [reply src/sparse_hessian.jl:20]: You have a `ncon == 0 ? xsi : [xsi; ysi]` and `ysi` is only used if it's defined.

File: src/sparse_jacobian.jl (hunk 1 above)
Comment (amontoison) [reply src/sparse_jacobian.jl:115]: I just want to harmonize the functions.

File: src/sparse_hessian.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+function hess_coord!(
+  b::SparseADHessian,
+  nlp::ADModel,
+  x::AbstractVector,
+  obj_weight::Real,
+  vals::AbstractVector,
+)
+  T = eltype(x)
+  y = zeros(T, nlp.meta.ncon)
```
[the remark this answers is not in the document]
Comment (amontoison) [reply src/sparse_hessian.jl:65]: Yes, we should also fix it in `NLPModelsJuMP.jl`.

File: src/sparse_hessian.jl [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    fun = fun + dot(c!(cx, xsi), ysi)
+  end
+  H = Symbolics.hessian_sparsity(fun, ncon == 0 ? xsi : [xsi; ysi], full = false)
+  H = ncon == 0 ? H : H[1:nvar,1:nvar]
+  rows, cols, _ = findnz(H)
+  vals = Symbolics.sparsehessian_vals(fun, xsi, rows, cols)
+  nnzh = length(vals)
+  # cfH is a Tuple{Expr, Expr}, cfH[2] is the in-place function
+  # that we need to update a vector `vals` with the nonzeros of ∇²ℓ(x, y, μ).
+  cfH = Symbolics.build_function(vals, xsi, ysi, μs, expression = Val{false})
```
Comment (tmigot) [inline src/sparse_hessian.jl:27]: but here `ysi` is not defined if `ncon = 0`.

File: src/sparse_hessian.jl [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  cfH::T
+end
+
+function SparseADHessian(nvar, f, ncon, c!; kwargs...)
+  @variables xs[1:nvar], μs
+  xsi = Symbolics.scalarize(xs)
+  fun = μs * f(xsi)
+  if ncon > 0
+    @variables ys[1:ncon]
+    ysi = Symbolics.scalarize(ys)
```
Comment (tmigot) [inline src/sparse_hessian.jl:16]:
```suggestion
  @variables ys[1:ncon]
  ysi = Symbolics.scalarize(ys)
  if ncon > 0
```

**Review by tmigot — CHANGES_REQUESTED** [review]
I still have the `ysi` not defined error.

File: src/sparse_hessian.jl (hunk 4 above)
Comment (amontoison) [reply src/sparse_hessian.jl:27]: I don't understand why the tests passed :thinking:

File: src/sparse_hessian.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (tmigot) [reply src/sparse_hessian.jl:65]: I guess the fastest solution is to store `y` in the `SparseADHessian` structure.

[the remark this answers is not in the document]
Comment (amontoison) [reply src/sparse_hessian.jl:65]: Do you want that I update the PR or open a new one after this one?

[the remark this answers is not in the document]
Comment (tmigot) [reply src/sparse_hessian.jl:65]: Let's do another one.

**Review by tmigot — APPROVED** [review]
Thanks @amontoison  !!!! #13 #44 #106

**Pull request merged**